src/core_modules/capture-core/components/DataEntries/EnrollmentRegistrationEntry/hooks/useBuildEnrollmentPayload.js

```diff
@@ -0,0 +1,189 @@
+    programId: string;
+    dataEntryId: string;
+    itemId?: string;
+    orgUnitId: string;
+    teiId: ?string;
+    trackedEntityTypeId: string;
+};
+
+function getClientValuesForFormData(formValues: Object, formFoundation: RenderFoundation) {
+    const clientValues = formFoundation.convertValues(formValues, convertFormToClient);
+    return clientValues;
+}
+
+function getServerValuesForMainValues(
+    values: Object,
+    meta: Object,
+    formFoundation: RenderFoundation,
+) {
+    const clientValues = convertDataEntryValuesToClientValues(
+        values,
+        meta,
+        formFoundation,
+    ) || {};
+
+    // potientally run this through a server to client converter for enrollment, the same way as for event
+    const serverValues = Object
+        .keys(clientValues)
+        .reduce((acc, key) => {
+            const value = clientValues[key];
+            const type = meta[key].type;
+            acc[key] = convertClientToServer(value, type);
+            return acc;
+        }, {});
+
+    return serverValues;
+}
+
+function getPossibleTetFeatureTypeKey(serverValues: Object) {
+    return Object
+        .keys(serverValues)
+        .find(key => key.startsWith('FEATURETYPE_'));
+}
+
+function buildGeometryProp(key: string, serverValues: Object) {
+    if (!serverValues[key]) {
+        return undefined;
+    }
+    const type = capitalizeFirstLetter(key.replace('FEATURETYPE_', '').toLocaleLowerCase());
+    return {
+        type,
+        coordinates: serverValues[key],
+    };
+}
+
+const geometryType = formValuesKey => Object.values(FEATURETYPE).find(geometryKey => geometryKey === formValuesKey);
+
+const deriveAttributesFromFormValues = (formValues = {}) =>
+    Object.keys(formValues)
+        .filter(key => !geometryType(key))
+        .map<{ attribute: string, value: ?any }>(key => ({ attribute: key, value: formValues[key] }));
+
+export const useBuildEnrollmentPayload = ({
+    programId,
+    dataEntryId,
+    itemId = 'newEnrollment',
+    orgUnitId,
+    teiId,
+    trackedEntityTypeId,
+}: DataEntryReduxConverterProps) => {
+    const dataEntryKey = getDataEntryKey(dataEntryId, itemId);
+    const formValues = useSelector(({ formsValues }) => formsValues[dataEntryKey]);
+    const dataEntryFieldValues = useSelector(({ dataEntriesFieldsValue }) => dataEntriesFieldsValue[dataEntryKey]);
+    const dataEntryFieldsMeta = useSelector(({ dataEntriesFieldsMeta }) => dataEntriesFieldsMeta[dataEntryKey]);
+    const { formFoundation: scopeFormFoundation } = useMetadataForRegistrationForm({ selectedScopeId: programId });
+    const { firstStageMetaData } = useBuildFirstStageRegistration(programId);
+    const { formFoundation } = useMergeFormFoundationsIfApplicable(scopeFormFoundation, firstStageMetaData);
+
+    const buildTeiWithEnrollment = (): EnrollmentPayload => {
+        // $FlowFixMe - Business logic dictates that formFoundation is not null in this callback
+        if (!formFoundation) return null;
```

Comment:
Thought about this a bit again, and I think it's better to just throw in situations where logic dictates some data to be there. This way we will not have any unintended consequences. 

```suggestion
	if (!formFoundation) throw Error('form foundation object not found');
```